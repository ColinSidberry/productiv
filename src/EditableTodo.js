import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * State: isBeingEdited: boolean
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [isBeingEdited, setIsBeingEdited] = useState(false);

  /** Toggle if this is being edited */

  // Goal flip is being edited, maybe safer call to just flip it, ternary
  // should never had side effects!!!!!!
  // Set to not edit  => (!edit)

  function toggleEdit() {
    isBeingEdited ? setIsBeingEdited(false) : setIsBeingEdited(true);
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    console.log(
      formData.id,
      formData.title,
      formData.description,
      "formdata from EditiableTodo.js"
    );
    update(formData);
    toggleEdit();
  }

  return (
    <div className="EditableTodo">
      {isBeingEdited ? (
        <TodoForm initialFormData={todo} createTodo={handleSave} />
      ) : (
        <div className="mb-3">
          <div className="float-right text-sm-right">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm"
              onClick={toggleEdit}
            >
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={handleDelete}
            >
              Del
            </button>
          </div>
          <Todo
            id={todo.id}
            title={todo.title}
            description={todo.description}
            priority={todo.priority}
          />
        </div>
      )}
    </div>
  );
}

export default EditableTodo;
