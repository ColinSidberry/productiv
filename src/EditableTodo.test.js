import { render, fireEvent } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

const TEST_DATA = {
  title: "test",
  description: "testing editable",
  priority: 1,
};

const update = jest.fn();
const remove = jest.fn();

describe("productiv editableTodo", function () {
  it("renders without crashing", function () {
    render(<EditableTodo todo={TEST_DATA} remove={remove} update={update} />);
  });

  it("matches snapshot", function () {
    const { container } = render(
      <EditableTodo todo={TEST_DATA} remove={remove} update={update} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders toggle edit button", function () {
    const { container } = render(<EditableTodo todo={TEST_DATA} />);

    expect(container).toContainHTML("test");
  });
});
