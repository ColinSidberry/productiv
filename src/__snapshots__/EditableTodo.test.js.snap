// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productiv editableTodo matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-right text-sm-right"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            test
          </b>
           
          <small>
            (priority: 
            1
            )
          </small>
        </div>
        <div>
          <small>
            testing editable
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
